src/Services/QueuedJobService.php

```diff
@@ -334,6 +300,98 @@ public function isAtMaxJobs()
         return false;
     }
 
+    /**
+     * Using a job signature, returns the JobDescriptor ID and whether the
+     * job descriptor is new or existing
+     *
+     * @param string $signature
+     * @param QueuedJob $job
+     * @param QueuedJobDescriptor $jobDescriptor
+     * @param null|string $startAfter
+     * @return int|null
+     */
+    protected function findOrMakeJobDescriptorFromSignature($signature, $job, $jobDescriptor, $startAfter)
+    {
+        // Start a transaction which will hold until we have a lock on this signature.
+        return DB::get_conn()->withTransaction(function () use ($signature, $job, $jobDescriptor, $startAfter) {
+            $query = 'SELECT "ID" FROM "QueuedJobDescriptor" WHERE "Signature" = ? FOR UPDATE';
```

Comment:
Bit of an edge case, though should read the tableName of QueuedJobDescriptor first incase it's been changed (table names are configuraable

`$tableName = Config::inst()->get(QueuedJobDescriptor::class, 'table_name', Config::UNINHERITED);`